crates/matrix-sdk-ui/src/authentication/oidc.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::HashMap,
+    fmt, fs,
+    fs::File,
+    io::{BufReader, BufWriter},
+    path::PathBuf,
+};
+
+use matrix_sdk::oidc::types::registration::{
+    ClientMetadata, ClientMetadataVerificationError, VerifiedClientMetadata,
+};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+#[derive(Debug, thiserror::Error)]
+pub enum OidcRegistrationsError {
+    #[error("Failed to use the supplied base path.")]
+    InvalidBasePath,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RegistrationDataError {
+    #[error("Failed to load registrations file: {message}")]
+    LoadFailure { message: String },
+    #[error("Metadata hash mismatch, clearing registrations")]
+    MetadataMismatch,
+}
+
+/// A client ID that has been registered with an OpenID Connect provider.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct ClientId(pub String);
+
+impl fmt::Display for ClientId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// The data needed to restore an OpenID Connect session.
+#[derive(Debug)]
+pub struct OidcRegistrations {
+    /// The path of the file where the registrations are stored.
+    file_path: PathBuf,
+    /// The hash for the metadata used to register the client.
+    /// This is used to check if the client needs to be re-registered.
+    verified_metadata: VerifiedClientMetadata,
+    /// Pre-configured registrations for use with issuers that don't support
+    /// dynamic client registration.
+    static_registrations: HashMap<Url, ClientId>,
+}
+
+/// The underlying data serialized into the registration file.
+#[derive(Debug, Serialize)]
+struct FrozenRegistrationData {
+    /// The hash for the metadata used to register the client.
+    metadata: VerifiedClientMetadata,
+    /// All of the registrations this client has made as a HashMap of issuer URL
+    /// (as a string) to client ID (as a string).
+    dynamic_registrations: HashMap<Url, ClientId>,
+}
+
+/// The deserialize data from the registration file. This data needs to be
+/// validated before it can be used.
+#[derive(Debug, Deserialize)]
+struct UnvalidatedRegistrationData {
```

Comment:
Out of curiosity, what's the risk assessment behind deserializing as unvalidated? to prevent against external apps writing invalid information? or in case the validation rules change, so as to invalidate previously valid registrations?

Reply:
I'm not exactly sure on this one. [ClientMetadata](https://github.com/matrix-org/matrix-authentication-service/blob/0bee1284b247272d3eb96c73eb9b1d92151f39a8/crates/oauth2-types/src/registration/mod.rs#L129-L131) is `Deserialize` only and [VerifiedClientMetadata](https://github.com/matrix-org/matrix-authentication-service/blob/0bee1284b247272d3eb96c73eb9b1d92151f39a8/crates/oauth2-types/src/registration/mod.rs#L806-L808) is `Serialize` only. @zecakeh or @sandhose can probably provide more context on the reasons for this.

Reply:
Ok, if that happens in a lib upstream, I suppose it's for the reasons I imagined. Curious to hear the reasoning, but let's not block merging for that.